library-best-practices/JLBP-2.md

```diff
@@ -2,19 +2,40 @@
 -----------------------------
 
 - Avoid exposing types from your dependencies.
-  - Your dependencies have their own rules about how often
-    they do major version bumps and make breaking changes.
-    Their frequency of such changes may be more often than yours,
-    which causes problems if you want to retain a stable surface for
-    longer than your dependencies. Historically this was a
-    big problem for Guava, which made a major version bump
-    every 6 months, although Guava no longer does this.
-- Use package-protected classes and methods when the API is internal
-  only and should not be used by consumers.
-- Do not mark methods and classes public by default.
-  Assume non-public until a need is known.
+
+  - Method arguments and return values from public methods should be standard Java 
+    types such as `java.time.LocalDate` or classes defined in the library itself,
+    not third party types such as `org.joda.time.LocalDate`.
+
+  - Your own public types should not be subclasses or implementations of types
+    in third party libraries.  
+
+  - Third party libraries can change their own API more frequently than you like
+    or do so at inconvenient times. If a third party type on the surface of
+    your API changes or is removed, you either have to break your own API or remain with an older,
+    unsupported version of the library that is likely to cause diamond dependency
+    problems. Historically this was a big problem for libraries that exposed Guava types
+    such as `com.google.common.io.OutputSupplier` because Guava bumped major versions
+    every 6 months.
+
+- Use package-protected classes and methods for internal APIs that should not be used by consumers.
+
+- Do not mark methods and classes public by default. Assume non-public until a need is 
+  known.<sup id='a1'>[1](#item15)</sup>
+
+- Design for inheritance or prohibit it. That is, mark classes final unless there is a clear
+  reason for them to be subclassed. Mark methods in non-final classes final unless they
+  are meant to be overridden.<sup id='a2'>[2](#item19)</sup>
+
 - Prefer fewer packages over more packages to avoid
   unnecessarily publicizing internal details,
-  since any dependency across package boundaries need to be
+  since any dependency across package boundaries needs to be
   public. (We may revisit this when we can rely on
   the new module system in Java 11 or later.) 
+
+
+<b id="item15">1</b> Bloch, Joshua. "Item 15: Minimize the accessibility of classes and members." Effective Java, 3rd Edition. Boston: Addison-Wesley, 2018. p. 73[↩](#a1)
+
+<b id="item19">2</b> Bloch, Joshua. "Item 19: Design and document for inheritance or else 
+prohibit it." Effective Java, 3rd Edition. Boston: Addison-Wesley, 2018. p. 93[↩](#a2)
```

Comment:
```suggestion
prohibit it." Effective Java, 3rd Edition. Boston: Addison-Wesley, 2017. p. 93[↩](#a2)
```

> Publication Date: December 18, 2017
https://www.amazon.com/Effective-Java-Joshua-Bloch-ebook/dp/B078H61SCH

Reply:
The book itself lists 2018. Publishers often bump the copyright year by +1 for books released late in the year. I *think* citations usually go by what's listed in the book.

Reply:
Thank you for clarification.